### PR TITLE
[Speed] feature/ParallelExecutor

### PR DESCRIPTION
**Pull request by tonyyang-svail**
fix https://github.com/PaddlePaddle/Paddle/issues/8592

Profiling result:
script: as the example in this pr
command:
```
CUDA_VISIBLE_DEVICES=0             nvprof -f -o one.nvvp python parallel_executor_example.py --batch_size=32
CUDA_VISIBLE_DEVICES=0,1,2,3       nvprof -f -o four.nvvp python  parallel_executor_example.py --batch_size=32
```
| Setting | copy weights | forward and backward | merge gradient | apply gradient |
|------|----------------|------------|----------------|----------------|
| 1  with nccl on bp  | /              | 250     | /              | 5              |
| 4 with nccl on bp   | /             | 750(AllReduce takes about 63%)     | /              | 5              |

### Save Model (to be implemented)

In the current implementation, the ParallelExecutor's constructor creates a base scope and *n* (*n* equals the number of GPUs) sub scopes, the model is replicated in each sub scope. The save model function **cannot** access the sub scopes created by the ParallelExecutor.

#### Proposed Solution

ParallelExecutor's constructor creates *n-1* sub scopes. `ParallelExecutor.run` will take a scope parameter, which will be attached as another sub scope. In this way, the user can create a scope, use it for `ParallelExecutor.run` as well as for save model.

### CODE REVIEW

File: python/paddle/fluid/tests/book/parallel_executor_example.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            conv_batchnorm_drop_rate=dropouts,
+            pool_type='max')
+
+    conv1 = conv_block(input, 64, 2, [0.3, 0])
+    conv2 = conv_block(conv1, 128, 2, [0.4, 0])
+    conv3 = conv_block(conv2, 256, 3, [0.4, 0.4, 0])
+    conv4 = conv_block(conv3, 512, 3, [0.4, 0.4, 0])
+    conv5 = conv_block(conv4, 512, 3, [0.4, 0.4, 0])
+
+    # drop = fluid.layers.dropout(x=conv5, dropout_prob=0.5)
```
Comment (helinwang) [inline python/paddle/fluid/tests/book/parallel_executor_example.py]: Could you remove the commented code?

File: python/paddle/fluid/parallel_executor.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from threading import Thread
+from Queue import Queue
+from core import CUDAPlace
+
+__all__ = ['ParallelExecutor']
+
+print_lock = Lock()
+
+
+def save_print(*args, **kwargs):
```
Comment (helinwang) [inline python/paddle/fluid/parallel_executor.py]: Is `save_print` and `pretty_id_indent` necessary? They are not used in this PR.

File: doc/design/parallel_executor.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+## Design
+
+#### API
+
+We don't expose `scope` as in the run interface because we use it to maintaining the inter-device
+variables. For example
+
+1. NCCL communicator
+1. Data reader(?)
```
Comment (helinwang) [inline doc/design/parallel_executor.md]: Do we need "?", if so probably need to explain it.

File: doc/design/parallel_executor.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+To train a neural network multi GPUs, a user only need to make the following modifications in
+the code
+1. specify `append_all_reduce` flag in the optimizer
+1. use `ParallelExecutor` to run the `programDesc`.
+
+```python
+cost = your_neural_network()
+
+opt = fluid.optimizer.SGDOptimizer(..., append_all_reduce=True)
```
Comment (panyx0718) [inline doc/design/parallel_executor.md]: what if append_all_reduce=True is not set, will it use other way to send gradients to each other?

File: python/paddle/fluid/optimizer.py [hunk 5]
```diff
@@ -35,7 +36,11 @@ class Optimizer(object):
     but need to use one of it's implementation.
     """
 
-    def __init__(self, learning_rate, regularization=None):
+    def __init__(self,
+                 learning_rate,
+                 global_step=None,
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: you might want to put global step next to avoid breaking exiting users?

File: python/paddle/fluid/optimizer.py [hunk 6]
```diff
@@ -35,7 +36,11 @@ class Optimizer(object):
     but need to use one of it's implementation.
     """
 
-    def __init__(self, learning_rate, regularization=None):
+    def __init__(self,
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: comments for the arguments?

File: python/paddle/fluid/optimizer.py [hunk 7]
```diff
@@ -53,6 +58,7 @@ def __init__(self, learning_rate, regularization=None):
         # {accum_name : { paramter_name : accumulator_for_parameter, ...}, ...}
         self._accumulators = defaultdict(lambda: dict())
         self.helper = None
+        self.append_all_reduce = append_all_reduce
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: why is this a public attribute?

File: python/paddle/fluid/parallel_executor.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        self.scope = core.Scope()
+        self.scopes = {}
+        for idx, _ in enumerate(self.executors):
+            self.scopes[idx] = self.scope.new_scope()
+
+        core.init_nccl_com(self.scope, gpu_list)
+
+    def run(self,
+            program=None,
```
Comment (panyx0718) [inline python/paddle/fluid/parallel_executor.py]: Is it possible to run part of the program in multi-gpus and keep others in 1 gpu or cpu

File: python/paddle/fluid/parallel_executor.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            feed=None,
```
Comment (panyx0718) [inline python/paddle/fluid/parallel_executor.py]: In the design doc, you said feed won't be added?

File: python/paddle/fluid/optimizer.py [hunk 10]
```diff
@@ -35,7 +36,11 @@ class Optimizer(object):
     but need to use one of it's implementation.
     """
 
-    def __init__(self, learning_rate, regularization=None):
+    def __init__(self,
+                 learning_rate,
+                 global_step=None,
+                 regularization=None,
+                 append_all_reduce=False):
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: I feel this to be an implementation detail. Can we avoid exposing it to user?

File: doc/design/parallel_executor.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+opt.minimize(avg_cost)
+
+exe = fluid.ParallelExecutor(gpu_list=[0, 1])
+```
+
+## Design
```
Comment (panyx0718) [inline doc/design/parallel_executor.md]: Briefly describe the plan for model save/restore?

File: python/paddle/fluid/parallel_executor.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            fetch_list=None,
+            feed_var_name='feed',
+            fetch_var_name='fetch',
+            return_numpy=True):
+        # TODO(helin): split input
+        q = Queue(maxsize=len(self.executors))
+        for idx, exe in enumerate(self.executors):
+            cur_scope = self.scopes[idx]
+            t = Thread(
```
Comment (panyx0718) [inline python/paddle/fluid/parallel_executor.py]: I'm a little worried about doing multi-threading in Python for such important computation, but maybe I'm wrong.

File: python/paddle/fluid/tests/book/parallel_executor_example.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        learning_rate=0.001, append_all_reduce=True)
+    opt.minimize(avg_cost)
+
+    # program_summary(fluid.default_main_program())
+
+    exe = fluid.ParallelExecutor(
+        gpu_list=range(fluid.core.get_cuda_device_count()))
+
+    # Parameter initialization
+    exe.run(fluid.default_startup_program())
```
Comment (panyx0718) [inline python/paddle/fluid/tests/book/parallel_executor_example.py]: So startup is also run in parallel?

File: python/paddle/fluid/parallel_executor.py (hunk 12 above)
Comment (typhoonzero) [reply python/paddle/fluid/parallel_executor.py]: I think need an `Executor` written in c++ and wrap it in python, so the thread in CPU can use multiple cores.

File: python/paddle/fluid/parallel_executor.py (hunk 8 above)
Comment (tonyyang-svail) [reply python/paddle/fluid/parallel_executor.py]: I don't think so.

File: python/paddle/fluid/tests/book/parallel_executor_example.py (hunk 13 above)
Comment (tonyyang-svail) [reply python/paddle/fluid/tests/book/parallel_executor_example.py]: Yes.

File: python/paddle/fluid/optimizer.py (hunk 7 above)
Comment (tonyyang-svail) [reply python/paddle/fluid/optimizer.py]: It should be private.

File: python/paddle/fluid/optimizer.py (hunk 10 above)
Comment (tonyyang-svail) [reply python/paddle/fluid/optimizer.py]: Sure.

File: doc/design/parallel_executor.md (hunk 4 above)
Comment (tonyyang-svail) [reply doc/design/parallel_executor.md]: Currently, it's not handled.

File: python/paddle/fluid/parallel_executor.py (hunk 12 above)
Comment (tonyyang-svail) [reply python/paddle/fluid/parallel_executor.py]: @typhoonzero Right. A design of C++ `ParallelExecutor` will be submitted in a separate PR.

**Comment by tonyyang-svail** [comment]
A c++ implementation of ParallelExecutor will be submitted at https://github.com/PaddlePaddle/Paddle/pull/9035.

This PR will be closed.

Because it would be relatively easier to adapt multi-stream at the C++ level.
